llvm/tools/ir2builder/ir2builder.cpp

```diff
@@ -0,0 +1,1702 @@
+          // since the basic blocks don't exist yet
+          OS << "// TODO: Basic block extraction\n";
+        }
+        initVal = getVal(G.getInitializer());
+      }
+
+      std::string gname = getVal(&G);
+      OS << "auto " << gname << " = new " << llvmPrefix << "GlobalVariable(*"
+         << modName << ", " << getType(G.getValueType()) << ", "
+         << to_str(G.isConstant()) << ", " << getLinkage(G) << ", " << initVal
+         << ", \"" << escape(G.getName().str()) << "\", "
+         << "nullptr"
+         << ", " << getThreadLocalMode(G.getThreadLocalMode()) << ", "
+         << std::to_string(G.getAddressSpace()) << ", "
+         << to_str(G.isExternallyInitialized()) << ");\n";
+    }
+  }
+
+  // Function generation
+  for (Function &F : M) {
+    if (!F.isDeclaration())
+      convert(F, OS);
+  }
+
+  // Epilogue
+  if (GenerateRunnable) {
+    // Print generated code
+    OS << "    " << modName << "->print(outs(), nullptr);\n";
+    // Cleanup
+    OS << "    delete " << modName << ";\n"
+       << "    return 0;\n}\n";
+  }
+}
+
+void IR2Builder::convert(Function &F, raw_ostream &OS) {
+  phiIncomings.clear();
+  // Function
+  OS << "{\n\n";
+  auto fDecl = getVal(&F);
+  OS << llvmPrefix << "Function *" << fDecl;
+  OS << " = " << llvmPrefix << "Function::Create("
+     << getType(F.getFunctionType()) << ", " << getLinkage(F) << ", \""
+     << F.getName() << "\", " << modName << ");\n";
+
+  OS << "\n";
+
+  // Set attributes
+  if (F.getCallingConv() != CallingConv::C) { // C is default
+    OS << fDecl << "->setCallingConv(" + getCallingConv(F.getCallingConv())
+       << ");\n";
+  }
+  // TODO: Handle attributes with values
+  if (F.getAttributes().hasFnAttrs()) {
+    for (auto att : F.getAttributes().getFnAttrs()) {
+      OS << fDecl << "->addFnAttr(";
+      outputAttr(att, OS);
+      OS << ");\n";
+    }
+  }
+  if (F.getAttributes().hasRetAttrs()) {
+    for (auto att : F.getAttributes().getRetAttrs()) {
+      OS << fDecl << "->addRetAttr(";
+      outputAttr(att, OS);
+      OS << ");\n";
+    }
+  }
+  for (size_t i = 0; i < F.arg_size(); ++i) {
+    if (F.getAttributes().hasParamAttrs(i)) {
+      for (auto att : F.getAttributes().getAttributes(i + 1)) {
+        OS << fDecl << "->addParamAttr(" << i << ", ";
+        outputAttr(att, OS);
+        OS << ");\n";
+      }
+    }
+  }
+
+  // Save arguments into variables for easy access
+  for (unsigned i = 0; i < F.arg_size(); ++i) {
+    OS << "auto " << getVal(F.getArg(i)) << " = " << fDecl << "->getArg("
+       << std::to_string(i) << ");\n";
+  }
+
+  if (UseRPO) {
+    ReversePostOrderTraversal<Function *> RPOT(&F);
+
+    // Basic block declaration in order
+    for (BasicBlock *BB : RPOT) {
+      std::string bbName = getVal(BB);
+      OS << llvmPrefix << "BasicBlock* " << bbName << " = " << llvmPrefix
+         << "BasicBlock::Create(" << ctx << ", \"" << BB->getName() << "\", "
+         << fDecl << ");\n";
```

Comment:
This whole patch could benefit from using LLVM's powerful `formatv` facility (https://llvm.org/docs/ProgrammersManual.html#formatv-examples) for string formatting. That goes for code like above using the stream insertion operator, or code elsewhere that cobbles `std::string`s together with "+". For example, the above can be written this way:

```
  OS << formatv("{0}BasicBlock* {1} = {0}BasicBlock::Create({2}, \"{3}\", {4});\n",
               llvmPrefix, bbName, ctx, BB->getName(), fDecl);
```
There are numerous other places this could apply.

Reply:
On a related note, the patch could also benefit from using LLVM's `Twine` and `StringRef` throughout.